### DISCUSSION TITLE
Wish list for next big features 🎉

### DISCUSSION THREAD

**Discussion by clue** [clue/framework-x#203] [Ideas]
## The next steps for Framework X 🎉

We want to bring Framework X forward and spend some more time on developing new features. While we've already started looking into some future topics, we’re especially interested about what our community would like to see (that's you!). This is why we're starting this discussion in the hopes to get more people involved and share their own ideas of what a future Framework X could look like.

## Suggest your ideas!

Bring in your ideas! Simply create a new comment thread in this discussion for each idea you want to see implemented in Framework X (any number of ideas, ideally one idea per comment thread).

We plan to use this discussion also to bring in our own ideas as well in a few days, but want to give more people a chance to suggest their ideas first. We're curious what our community wants and how our plans align!

## Help us prioritize

Everybody is invited to share their ideas! Some ideas will be better, some ideas will be less obvious. Help us understand what you need most.

You can cast your voice by voting on each comment thread or leave feedback in a comment thread if suggestions need more context. If a suggestion requires more discussion space, we'll try to create a separate discussion and link from here to create a common overview.

We're equally interested in feedback for your ideas and our own ideas and plan to adjust our roadmap according to the needs of our users.

## Next steps

Share ideas, leave feedback, vote. Depending on feedback, we plan to keep this open for a couple of weeks and will then announce further plans.

We're recently started a similar discussion to help shape ReactPHP `vNEXT` (https://github.com/orgs/reactphp/discussions/472) and we're excited to see how we could similarly shape the future of Framework X together.

Looking forward to your ideas and feedback! ❤️

**Comment by pfk84** [comment]
Proper upload handling (not limited to 64KB)/streaming requests and/or configuration of internal webserver, however you name or wish to implement it... Definitively the highest priority for me.

**Reply by clue** [reply]
Agree that this is one of the bigger limitations of using the built-in web server at the moment. As discussed in #69 and #160, there are a number of workarounds and possible patches, but a full solution would indeed require a bigger effort. Curious how this will be voted on compared to the other ideas!

**Comment by pfk84** [comment]
PSR-15 compatibility so we can easily add existing middleware like CORS, compression etc.

**Reply by JustSteveKing** [reply]
PSR-15 is a gold standard now for micro-frameworks, I would love to see this implemented. Especially if we can lean on the DI container too!

```php
$app->get(
    '/',
    MyController::class,
    [
        SomeMiddleware::class,
        AnotherMiddleware::class
    ]
);
```

**Reply by clue** [reply]
Agree that we absolutely want to have this in Framework X for broad interoperability! As discussed in #130 and #185, it's not too hard to use custom adapters, but I agree that we should support this out of the box in the future. This would also neatly align with the plans for ReactPHP v3 getting more fiber-based APIs as discussed in https://github.com/orgs/reactphp/discussions/472.

**Reply by JustSteveKing** [reply]
Async middleware would also be an interesting concept ..... Just saying

**Reply by SimonFrings** [reply]
@JustSteveKing This is actually already supported, but we still need to write some more documentation for this :D

**Comment by pfk84** [comment]
Error details configuration option for the built-in `ErrorHandler` to be able to surpress Exception details for the public eye.

**Reply by clue** [reply]
Agree that we should provide more control over this behavior as discussed in #170! Thanks to the preparations in #173 and #175 this should now already be possible using a custom middleware, but I agree this should be included out of the box in the future once we see more demand.

**Comment by SimonFrings** [comment]
I would really like to see a first (beta) release for Framework X and a tagged version. I think this will bring the project forward and more people feel safe to install the project.

**Comment by SimonFrings** [comment]
I think we should work out the remaining points in the [documentation](https://framework-x.org/docs/getting-started/quickstart/) which are still under construction. I don't think it's realistic to do every point in one go, but maybe figure out what the most important ones are and start there.

**Comment by geertvanbommel** [comment]
Websocket support similar to kemalcr.com
```
require "kemal"

# Matches GET "http://host:port/"
get "/" do
  "Hello World!"
end

# Creates a WebSocket handler.
# Matches "ws://host:port/socket"
ws "/socket" do |socket|
  socket.send "Hello from Kemal!"
end

Kemal.run
```

**Reply by SimonFrings** [reply]
I agree, definitely something I can see on the roadmap for Framework X.  This is already currently possible as described in https://github.com/clue/framework-x/discussions/85, but a built-in functionality would be more comfortable :+1:
